### PR TITLE
iss18&19

### PR DESCRIPTION
**Pull request by qxp134**

### CODE REVIEW

File: 18.cpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+#include <algorithm>
+#include <iostream>
+#include <vector>
+
+class Solution {
+ public:
+  std::vector<std::vector<int> > fourSum(
+      const std::vector<int>& nums, int target) {
+    std::vector<std::vector<int> > four_sum_vec;
+    std::sort(nums.begin(), nums.end());
```
Comment (chenleexyz) [inline 18.cpp:11]: This line should not compiles. use `std::vector<int>&`

File: 18.cpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    int last_fri_num = 999;
```
Comment (chenleexyz) [inline 18.cpp]: `fri` means `first`? Then `1st`, `2nd` is shorter and cleaner.

File: 18.cpp [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+          if (diff > 0) {
+            --l;
+          } else if (diff < 0) {
+            ++k;
+          } else {
+            base_vec.push_back(nums[i]);
+            base_vec.push_back(nums[j]);
+            base_vec.push_back(nums[k]);
+            base_vec.push_back(nums[l]);
+            four_sum_vec.push_back(base_vec);
```
Comment (chenleexyz) [inline 18.cpp]: `four_sum_vec.emplace_back(std::vector<int>({nums[i], nums[j], nums[k], nums[l]}));`

File: 18.cpp (hunk 2 above)
Comment (chenleexyz) [reply 18.cpp]: BTW, why do you use 999?

File: 19.cpp [hunk 4]
```diff
@@ -0,0 +1,30 @@
+#include <iostream>
+
+struct ListNode {
+  int val;
+  ListNode* next;
+  ListNode(int x) : val(x), next(NULL) {}
```
Comment (chenleexyz) [inline 19.cpp:6]: prefer `nullptr` to NULL.

File: 19.cpp [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+class Solution {
+ public:
+  ListNode* removeNthFromEnd(ListNode* head, int n) {
+    ListNode* frist_ptr = head;   // found the end node
+    ListNode* second_ptr = head;  // a node before will be deleted node
+    while (n != 0) {
+      --n;
+      frist_ptr = frist_ptr->next;
+    }
+    if (frist_ptr == NULL)
```
Comment (chenleexyz) [inline 19.cpp]: I think `if (first_ptr)` is clear enough.

File: 19.cpp [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+struct ListNode {
+  int val;
+  ListNode* next;
+  ListNode(int x) : val(x), next(NULL) {}
+};
+class Solution {
+ public:
+  ListNode* removeNthFromEnd(ListNode* head, int n) {
+    ListNode* frist_ptr = head;   // found the end node
+    ListNode* second_ptr = head;  // a node before will be deleted node
```
Comment (chenleexyz) [inline 19.cpp]: I think name `fast_ptr`, `slow_ptr` is more intuitive.

File: 19.cpp [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    while (n != 0) {
+      --n;
+      frist_ptr = frist_ptr->next;
+    }
+    if (frist_ptr == NULL)
+      return head->next;
+    while (frist_ptr->next != NULL) {
```
Comment (chenleexyz) [inline 19.cpp]: ditto.

File: 19.cpp [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+      frist_ptr = frist_ptr->next;
+      second_ptr = second_ptr->next;
+    }
+    ListNode* delete_ptr = second_ptr->next;
+    if (delete_ptr != NULL) {
+      second_ptr->next = delete_ptr->next;
+      delete (delete_ptr);
+    }
+    return head;
```
Comment (chenleexyz) [inline 19.cpp:30]: It seems `remove(1->2, 2)` will fail?

File: 19.cpp [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+      frist_ptr = frist_ptr->next;
+    }
+    if (frist_ptr == NULL)
+      return head->next;
+    while (frist_ptr->next != NULL) {
+      frist_ptr = frist_ptr->next;
+      second_ptr = second_ptr->next;
+    }
+    ListNode* delete_ptr = second_ptr->next;
+    if (delete_ptr != NULL) {
```
Comment (chenleexyz) [inline 19.cpp]: ditto.

File: 19.cpp (hunk 8 above)
Comment (qxp134) [reply 19.cpp:30]: it works,will return head->next

File: 19.cpp (hunk 4 above)
Comment (qxp134) [reply 19.cpp:6]: leetcode give this struct

Comment (chenleexyz) [reply 19.cpp:6]: I see.

File: 19.cpp (hunk 8 above)
Comment (chenleexyz) [reply 19.cpp:30]: Fine.

**Comment by chenleexyz** [comment]
LGTM

File: 19.cpp [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+ public:
+  ListNode* removeNthFromEnd(ListNode* head, int n) {
+    ListNode* frist_ptr = head;   // found the end node
+    ListNode* second_ptr = head;  // a node before will be deleted node
+    while (n != 0) {
+      --n;
+      frist_ptr = frist_ptr->next;
+    }
+    if (frist_ptr == NULL)
+      return head->next;
```
Comment (chenleexyz) [inline 19.cpp]: I this case, `head` is not deleted.

File: 19.cpp (hunk 5 above)
Comment (qxp134) [reply 19.cpp]: use this  if (!first_ptr)

**Pull request merged**